cylc/flow/task_pool.py

```diff
@@ -1477,304 +1450,530 @@ def can_spawn(self, name: str, point: 'PointBase') -> bool:
             LOG.debug(
                 'Not spawning %s/%s: before initial cycle point', point, name)
             return False
-        elif self.config.final_point and point > self.config.final_point:
+
+        if self.config.final_point and point > self.config.final_point:
             # Only happens on manual trigger beyond FCP
             LOG.debug(
                 'Not spawning %s/%s: beyond final cycle point', point, name)
             return False
+
+        # Is it on-sequence and within recurrence bounds.
+        if not self.config.get_taskdef(name).is_valid_point(point):
+            LOG.warning(
+                self.ERR_PREFIX_TASK_NOT_ON_SEQUENCE.format(
+                    name, point
+                )
+            )
+            return False
+
         return True
 
+    def _get_task_history(self, name, point, flow_nums):
+        """Get details of previous submits for this task.
+
+        """
```

Comment:
The task history code has been re-written here so I tried replicating this https://github.com/cylc/cylc-flow/issues/5952 on this branch.

Unfortunately it did replicate, however, I'm happy for this to be solved in a later PR, but I did notice some other behaviour in the process which looked a bit fishy.

In the example, the task `4/a` is run, the scheduler is then restarted and the workflow stalls because a task which depends on `4/a` erroneously claims this task has not generated the succeeded output.

Interesting, when I jam logging in on this branch, I can see that no attempt was made to load `4/a` from the DB (likely as a result of a recurrence change?), however, some unexpected attempts are made to load in tasks in later cycles.

```python
# $ cylc play <workflow> -s stop_cycle=6 -N --debug
# LOADING task proxies
>>> _get_task_history(a, 6, {1})  # why are we looking up 6/a and 7/a but not 5/a or 4/a?
(0, False, False)
>>> _get_task_history(a, 7, {1})  # why are we looking up tasks beyond the stop cycle point?!
(0, False, False)
# ...
# [5/a/01:running] => succeeded
>>> _get_task_history(b, 5, {1})  # eh? why are we looking up the task that just completed?!
(0, False, False)
# [6/a/01:running] => succeeded
>>> _get_task_history(b, 6, {1})  # eh? any why are we checking this again? It's been in the pool since the last check!
(0, False, False)
```

Reply:
> The task history code has been re-written here so I tried replicating this https://github.com/cylc/cylc-flow/issues/5952 on this branch since the logic has been re-written.
> 
> Unfortunately it did replicate,

Yeah that's off-topic for this branch - see my comments on the issue.

Reply:
> In the example, the task 4/a is run, the scheduler is then restarted and the workflow stalls because a task which depends on 4/a erroneously claims this task has not generated the succeeded output.

It's not claiming that, although the stall message unfortunately looks like it - see the other issue.

Reply:
Your `_get_task_history()` questions are kind of specific to this branch though:

- `# why are we looking up 6/a and 7/a but not 5/a or 4/a?`

Because we look up history when a new task is spawned, to get submit number and to see if was ever spawned before in this flow.  That explains `6/a` and `7/a`.  For the others, `4/a` is a past completed task, we're not spawning it again; and `5/a` was spawned before the restart; afterward it is loaded from the DB.

> `# why are we looking up tasks [7/a] beyond the stop cycle point?!` 

Because it's only a stop cycle point, not the final cycle point (i.e., it's not the end of the graph). The workflow will shutdown with `7/a` waiting in the task pool, to allow a restart to carry on after the old stop point.

>  # eh? why are we looking up [5/b] the task that just completed?!

From your transcript above it looks like `5/a` just completed, not `5/b` - right?

> `# eh? any why are we checking this [6/b] again? It's been in the pool since the last check!`

Ah, again I think you're neglecting the fact that `a` and `b` are different tasks??

Reply:
Summary: 
- we're punting what do about the modified-graph restart to #5952 
- I don't think there are any other valid concerns here

Plz resolve this conversation if you agree!